os/fs/driver/mtd/smart.c

```diff
@@ -3430,8 +3430,25 @@ static uint16_t smart_findfreephyssector(FAR struct smart_struct_s *dev, uint8_t
 				/* Assign this block to alloc from. */
 
 				if (x < dev->neraseblocks - 1 || !allocfreecount) {
-					allocblock = block;
-					allocfreecount = count;
+					int sector;
```

Comment:
IMO, you have to clean up this & current code
As you know, flow is below.
1. Check all erase block and check count of free-sector of each block.
2. If there's free sector in specific block, check wear level.
 2-1. if wearlevel is acceptable, checked block becomes a target block
3. If there's no free sector after flow #2, relocate block & retry #1~#2
4. Now we found target block, then check all sector header in block.
5. If it found sector which has clean header, then verify that contents area's data.
  5-1 if any of contents is flipped, update erase count and something, then retry it
6. If any of sector is all cleaned, then allocate it, otherwise it is program bug.

So your approach is to prevent #6 because of invalid header, I think, am I right?
Then this is not efficient code, because we already check header's status in #6. 
I think you can modify something in below.
```
		if ((UINT8TOUINT16(header.logicalsector) == 0xFFFF) &&
#if SMART_STATUS_VERSION == 1
			((header.seq == 0xFF) && (header.crc8 == 0xFF)) &&
#else
			(header.seq == CONFIG_SMARTFS_ERASEDSTATE) &&
#endif
			(!(SECTOR_IS_COMMITTED(header)))) {
```
This can be modify to HEADER_IS_CLEAN but HEADER_IS_CLEAN doesn't care CONFIG_SMARTFS_ERASEDSTATE, you have to modify it too.
```
```
if (HEADER_IS_CLEAN(header)) {
    check_bitflip_in_contents;
} else {
    compare_with_sMap;
    if (sMap is valid but header is not clean) {
          release_sector;
          update_freecount & release_count;
   }
   ...
}
```

Reply:
Also if allocation falied because of flipped bit, retry is necessary for both of if-else in my comment.